### ISSUE TITLE
Long file paths crash compilation on windows

### ISSUE THREAD

**Issue by WebFreak001** [dlang/dub#911]
When I compile my dub package in a nested folder somewhere in like `C:\Users\MyUsername\Development\My_D_Projects\From2016\Old\my-super-awesome-project-with-some-long-filename` and I have some dependency in it like `experimental_allocator` it will crash with this error message:

```
Error: Error writing file '..\..\..\..\Roaming\dub\packages\experimental_allocator-2.70.0-b1\experimental_allocator\.dub\build\library-release-windows-x86-dmd_2071-09237FFF1A90ED11A7976B16E7AB1A89\experimental_allocator.lib'
dmd failed with exit code 1.
```

On Windows there is by default a file path length limit of 259 characters and either dub or dmd aren't using the extended windows paths using `\\?\` which increases the length limit to 32767 characters. So Windows doesnt allow you to write the file.

Reproducable project in my case:

``` bat
cd %appdata%
cd ..
cd Local\Temp
mkdir workspaced-install-636070322740907409
cd workspaced-install-636070322740907409
git clone https://github.com/Hackerpilot/DCD.git DCD
cd DCD
dub build --build=release --config=client --force
```

crash

**Comment by WebFreak001** [comment]
Another issue which probably has the same cause:
```
libdparse ~master: building configuration "library"...
Error: failed to create path to file: .dub\obj\libdparse\.dub\build\library-debug-windows-x86_64-ldc_0-DCDD4FDB2EDFDAE1509F1E37E3AAEA55\dparse.lib
no such file or directory
ldc2 failed with exit code 1.
```

**Comment by WebFreak001** [comment]
Workaround for Windows 10: in the registry set `HKLM\SYSTEM\CurrentControlSet\Control\FileSystem LongPathsEnabled` to 1 and then reboot or restart the command line and editor

**Comment by helgaros** [comment]
I think it's useful to use the long path tool program

**Comment by s-ludwig** [comment]
Optlink can surely be added to the list, too. I think the DUB fix for this should happen in Phobos. There is no reason to let this ugly piece of WinAPI design leak through to all application code in a cross platform environment.

**Comment by s-ludwig** [comment]
Just mentioning that a tiny improvement to this has been merged: #1242

**Comment by WebFreak001** [comment]
fixed by #2589
